## 1.Importing Libraries

In [515]:
import os
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

## 2. Reading the data

In [518]:
flight=pd.read_csv("flight_price.csv")
flight


Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [520]:
flight.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


## 3. Preliminary Analysis

3.1 check data type

In [524]:
flight.head(2)

Airline Date_of_Journey    Source Destination                  Route  \
0     IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1  Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662

In [526]:
flight.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

3.2 check for dublicates

In [534]:
flight.duplicated().sum()

220

In [536]:
(
    flight
    .loc[flight.duplicated(keep=False)]
    .sort_values(["Airline","Date_of_Journey","Source","Destination"])
)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

3.3 observation

-the type of 'date of journey','departure time' and 'arrival time' should change to  datetime
-the type of 'duration' and 'total_stops' is mixed, it should be numeric value(for 2h 50min , it should be 170 min)
-there are 220 dublicates and it should be removed


## 4.Detailed analysis

### Airlines

In [543]:
flight.Airline

0             IndiGo
1          Air India
2        Jet Airways
3             IndiGo
4             IndiGo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

In [545]:
flight.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

-some of the entries have inconsistant values

# Clean the 'Airline' column by replacing " Premium economy" with an empty string
# Then get the unique values from the cleaned column
(
   flight
    .Airline
    .str.replace(" Premium economy", "")
    .str.replace(" Business", "")
    .str.title()
    .unique()
)




### Date of journey

In [550]:
flight.Date_of_Journey

0        24/03/2019
1         1/05/2019
2         9/06/2019
3        12/05/2019
4        01/03/2019
            ...    
10678     9/04/2019
10679    27/04/2019
10680    27/04/2019
10681    01/03/2019
10682     9/05/2019
Name: Date_of_Journey, Length: 10683, dtype: object

-dtype is object so, we need to handle this

In [553]:
pd.to_datetime(flight['Date_of_Journey'], format='%d/%m/%Y')

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

### source

In [556]:
flight.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

### Destination

In [559]:
flight.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

### Route

-we need to drop this feature because route and stops give same information

### dep_time

In [564]:
## method-1 flight.Dep_Time.unique()
## method 2:

(
    flight
    .Dep_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")]
)

Series([], Name: Dep_Time, dtype: object)

In [566]:
pd.to_datetime(flight.Dep_Time).dt.time

C:\Users\Rahul-PC\AppData\Local\Temp\ipykernel_16092\2899660464.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(flight.Dep_Time).dt.time


0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

### arrival time

In [569]:
(
   flight
    .Arrival_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")]
    .str.split(" ").str[0]
)

0        01:10
2        04:25
6        10:25
7        05:05
8        10:25
         ...  
10666    19:00
10667    20:20
10672    19:00
10673    04:25
10674    21:20
Name: Arrival_Time, Length: 4335, dtype: object

### duration

In [572]:
flight.Duration

0        2h 50m
1        7h 25m
2           19h
3        5h 25m
4        4h 45m
          ...  
10678    2h 30m
10679    2h 35m
10680        3h
10681    2h 40m
10682    8h 20m
Name: Duration, Length: 10683, dtype: object

In [574]:
(
    flight
    .Duration
    .loc[lambda ser: ~ser.str.contains("m")]
    .unique()
)

array(['19h', '23h', '22h', '12h', '3h', '5h', '10h', '18h', '24h', '15h',
       '16h', '8h', '14h', '20h', '13h', '11h', '9h', '27h', '26h', '4h',
       '7h', '30h', '21h', '28h', '47h', '6h', '25h', '38h', '34h'],
      dtype=object)

In [576]:
(
    flight
    .Duration
    .loc[lambda ser: ~ser.str.contains("h")]
)

6474    5m
Name: Duration, dtype: object

In [578]:
flight.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

-the observation indexed 6474 is not valid as it has duration of 5 minutes which is clearly wrong.so we need to drop this row

In [581]:
(
    flight
    .Duration
    .drop(index=6474)
    .str.split(" ",expand=True) ## expand=True will give dataframe
    .set_axis(["hour","minute"],axis=1) # change the name of columns
    .assign(
        hour=lambda df_:(
            df_
            .hour
            .str.replace("h","")
            .astype(int)
            .mul(60)
        )
        ,
        minute=lambda df_:(
            df_.minute
            .str.replace("m","")
            .fillna("0")
            .astype(int)
        )
    )
    .sum(axis=1) ##This will sum the values for each row  as axis =1
    #.dtypes
    #.isna().sum()
    
)

0         170
1         445
2        1140
3         325
4         285
         ... 
10678     150
10679     155
10680     180
10681     160
10682     500
Length: 10682, dtype: int64

## total stops

In [584]:
flight.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [586]:
(
    flight
    .Total_Stops
    .replace("non-stop","0")
    .str.replace(" stops?", "", regex=True) # " stop" " stops"
    #.astype(int) is not work here as there is mising value in total_stops columns ans nan is float bydefaultt
    .pipe(lambda ser: pd.to_numeric(ser)) # it compatible with float beacuse nan is also float so, it give output in float format
    
    .unique()
    
)

array([ 0.,  2.,  1.,  3., nan,  4.])

## additional info

In [589]:
flight.Additional_Info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

## 5.Cleaning operation

In [592]:
def convert_to_minutes(ser):
    return(
        ser
            .str.split(" ",expand=True) ## expand=True will give dataframe
            .set_axis(["hour","minute"],axis=1) # change the name of columns
            .assign(
             hour=lambda df_:(
                df_
                .hour
                .str.replace("h","")
                .astype(int)
                .mul(60)
               )
              ,
              minute=lambda df_:(
                df_.minute
                .str.replace("m","")
                .fillna("0")
                .astype(int)
               )
              )
              .sum(axis=1) ##This will sum the values for each row  as axis =1
              )

In [594]:
def clean_data(df):
    return (
        df
        .drop(index=[6474])
        .drop_duplicates()
        .assign(
            #airline=lambda df_:df_.airline.str.lower()
             **{
                 col: df[col].str.strip()
             for col in df.select_dtypes(include="O").columns
             }
           
        )
        .rename(columns=str.lower)
        .assign(
            airline=lambda df_:(
                df_
                .airline
                .str.replace(" Premium economy", "")
                .str.replace(" Business", "")
                .str.title()
            ),
            date_of_journey=lambda df_: pd.to_datetime(df_.date_of_journey,format='%d/%m/%Y')
            ,
            destination= lambda df_:(
                df_
                .destination
                .str.replace("Banglore","Bangalore")
                .str.replace("Cochin", "Kochi")
            ),
            dep_time=lambda df_:pd.to_datetime(df_.dep_time).dt.time,
            #arrival_time = lambda df_: df_.arrival_time.loc[lambda ser: ser.str.contains("[^0-9:]")]
            #.str.split(" ").str[0]
            arrival_time=lambda df_:pd.to_datetime(df_.arrival_time).dt.time,
            #method-1---> duration=lambda df_: convert_to_minutes(df_.duration)
            #method-2:
            duration=lambda df_:df_.duration.pipe(convert_to_minutes) # here we simply pass the df_.duration columns to "convert_to_minutes" function
            ,
            total_stops=lambda df_:(
                df_
                .total_stops
                .replace("non-stop","0")
                .str.replace(" stops?", "", regex=True) # " stop" " stops"
                #.astype(int) is not work here as there is mising value in total_stops columns ans nan is float bydefaultt
                .pipe(lambda ser: pd.to_numeric(ser)) # it compatible with float beacuse nan is also float so, it give output in float format
        
            )
            ,
            additional_info=lambda df_: df_.additional_info.replace("No info","No Info")
        
            
        )
        .drop(columns="route")
        
        
       
    )

In [596]:
clean_data(flight).additional_info.unique()


C:\Users\Rahul-PC\AppData\Local\Temp\ipykernel_16092\567379650.py:31: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time=lambda df_:pd.to_datetime(df_.dep_time).dt.time,
C:\Users\Rahul-PC\AppData\Local\Temp\ipykernel_16092\567379650.py:34: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time=lambda df_:pd.to_datetime(df_.arrival_time).dt.time,


array(['No Info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

In [597]:
flight_cleaned = clean_data(flight)
flight_cleaned

C:\Users\Rahul-PC\AppData\Local\Temp\ipykernel_16092\567379650.py:31: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time=lambda df_:pd.to_datetime(df_.dep_time).dt.time,
C:\Users\Rahul-PC\AppData\Local\Temp\ipykernel_16092\567379650.py:34: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time=lambda df_:pd.to_datetime(df_.arrival_time).dt.time,


airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata   Bangalore  05:50:00   
2      Jet Airways      2019-06-09     Delhi       Kochi  09:25:00   
3           Indigo      2019-05-12   Kolkata   Bangalore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata   Bangalore  19:55:00   
10679    Air India      2019-04-27   Kolkata   Bangalore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi       Kochi  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No Info   3897  
1         13:15:00       445          2.0         No Info   7662  
2         04:25:00      1140          2.0         No Info  13882  
3         23:30:00       325          1.0         No Info   6218  
4         21:35:00       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No Info   4107  
10679     23:20:00       155          0.0         No Info   4145  
10680     11:20:00       180          0.0         No Info   7229  
10681     14:10:00       160          0.0         No Info  12648  
10682     19:15:00       500          2.0         No Info  11753  

[10462 rows x 10 columns]

## 6.split the Data

In [601]:
flight_final=flight_cleaned.sample(1000)

In [613]:
x=flight_final.drop(columns="price")
y=flight_final.price.copy()

print(x.shape,y.shape)


(1000, 9) (1000,)


In [611]:
x_,x_test,y_,y_test = train_test_split(x,y,test_size=0.2, random_state=42)
x_train,x_val,y_train,y_val=train_test_split(x_,y_,test_size=0.2,random_state=42)

print(x_.shape,y_.shape)
print(x_train.shape,y_train.shape)
print(x_val.shape,y_val.shape)
print(x_test.shape,y_test.shape)

(800, 9) (800,)
(640, 9) (640,)
(160, 9) (160,)
(200, 9) (200,)


## 7.Export the subsets

In [624]:
## need to improve this code as this code not work. i add train, test, and validation file manually in folder

# Function to export X (features) and y (labels) as a CSV file
def export_data(X, y, name):
    # Specify the directory where you want to save the file
    # For local Jupyter directory, save in the current directory or a subdirectory (like 'data/')
    local_directory = 'data'  # Change to '' if you want to save in the current directory

    # Check if the directory exists, if not, create it
    if not os.path.exists(local_directory):
        os.makedirs(local_directory)

    # Create the file name and the full path
    file_name = f"{name}.csv"
    file_path = os.path.join(local_directory, file_name)

    # Join the feature and label DataFrames/Series and export to CSV
    X.join(y).to_csv(file_path, index=False)
    
    # Return the first few rows of the saved CSV file for confirmation
    return pd.read_csv(file_path).head()
    